tests/publiccloud/xfsprepare.pm

```diff
@@ -117,34 +117,63 @@ sub create_config {
     assert_script_run("echo 'MKFS_OPTIONS=\"$mkfs_options\"' >> $CONFIG_FILE") unless ($mkfs_options eq '');
 }
 
-# Get all disks and partitions and return the first device, which has no parititons
-sub get_unused_nvme_device {
+# Get all disks and partitions and return the first device, which has no partitons
+# If the $dev_size argument is given, require the device to match the size aswell
+sub get_unused_device {
+    my ($dev_size) = @_;    # in GB
     my @disks = split /\n/, script_output('lsblk | grep disk | cut -d " " -f1');
+    my @sizes = split /\n/, script_output("lsblk -b | grep disk | awk '{print \$4}'");
     my @parts = split /\n/, script_output('lsblk | grep part | cut -d " " -f1');
-    foreach my $disk (@disks) {
-        # smartmatch is still experimental, but this can be enabled in the near future.
-        #return "/dev/$disk" unless ($disk ~~ @parts);
-        return "/dev/$disk" unless (grep { m/$disk/ } @parts);
+    for (my $i = 0; $i < @disks; $i++) {
+        my $disk = $disks[$i];
+        my $size = $sizes[$i] / (1024**3);    # in GB
+        unless (grep { m/$disk/ } @parts) {
+            if (defined($dev_size)) {
+                my $size_matches = sprintf("%.1f", $size) == sprintf("%.1f", $dev_size);
+                return "/dev/$disk" if $size_matches;
+            } else {
+                return "/dev/$disk";
+            }
```

Comment:
Here you repeatedly test `defined...`.
IMHO, is simpler refactoring this in `if defined dev_size` and `else` braches.